CMPUT301F23T20/JavaJeniuses
# Scanning Features Complete - ijoffe

The functionality behind user stories 5.1 and 5.2 are complete. They are also all documented, but I will add UI tests later (they will probably be pretty difficult).

I tested on this serial number:
![Screenshot 2023-11-26 103056](https://github.com/CMPUT301F23T20/JavaJeniuses/assets/90410161/10b5c3ce-21e9-4286-a60e-43ef2c5f83c7)

And on this barcode:
![Screenshot 2023-11-26 104503](https://github.com/CMPUT301F23T20/JavaJeniuses/assets/90410161/09f93e5d-0b4f-4614-bca8-2f9988325601)

To test it:
* You will have to set your device camera to be VirtualScene (Device Manager -> Pencil -> Show Advanced Settings -> Back -> VirtualScene).
* Then you will have to open the emulator inside another window (File -> Settings -> Emulator -> Deselect "Launch in a Tool Window").
* Then, once running, upload these or any sample images (Three dots -> Camera -> Upload images onto wall and table).
* Then, click one of the scan buttons in the add/edit fragments to open the Camera and the virtual environment.
* Walk into the room (hold down ALT, then Mouse is looking, W is forward, S backwards, D right, A left, Q down, and E up).
* Take a good picture of the number/barcode and it should work.

To make it work, I used Google ML's [barcode scanning API](https://developers.google.com/ml-kit/vision/barcode-scanning/android) and [text reading API](https://developers.google.com/ml-kit/vision/text-recognition/v2/android). For fetching info from barcodes, I used [this](https://www.barcodelookup.com/api) public API on the free trial, which should last until past the due date and we can use someone else's email to make a new one if needed.

Please review and let me know if there are any issues or if I should make any changes.

### Important
@donchuru and @KareemAssaf -- after taking a look, I believe this implementation will not be directly compatible with your photos branch. There will be merge conflicts and also deeper issues because we both rely on the same `onActivityResult()` method. Please let me know when you open a PR -- I know how to resolve these merge conflicts and make sure both of ours work at once, I think I have a good idea of how to make these coexist.
